pkg/interop/client.go

```diff
@@ -0,0 +1,136 @@
+	"bytes"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"reflect"
+	"strings"
+	"time"
+
+	"go.thethings.network/lorawan-stack/pkg/encoding/lorawan"
+	"go.thethings.network/lorawan-stack/pkg/ttnpb"
+	"go.thethings.network/lorawan-stack/pkg/types"
+)
+
+const (
+	// loRaAllianceDomain is the domain of LoRa Alliance.
+	loRaAllianceDomain = "lora-alliance.org"
+
+	// loRaAllianceJoinEUIDomain is the LoRa Alliance domain used for JoinEUI resolution.
+	loRaAllianceJoinEUIDomain = "joineuis." + loRaAllianceDomain
+
+	// loRaAllianceNetIDDomain is the LoRa Alliance domain used for NetID resolution.
+	loRaAllianceNetIDDomain = "netids." + loRaAllianceDomain
+)
+
+func LoRaAllianceJoinServerFQDN(eui types.EUI64) string {
+	o := []rune(strings.ToLower(eui.String()))
+	r := make([]rune, 2*len(o)-1)
+	// Inspired by https://github.com/golang/example/blob/46695d81d1fae905a270fb7db8a4d11a334562fe/stringutil/reverse.go#L20-L27
+	for i, j := 0, len(o)-1; i < len(o)/2; i, j = i+1, j-1 {
+		r[i+i], r[j+j] = o[j], o[i]
+		r[i+i+1] = '.'
+		r[j+j-1] = '.'
+	}
+	return fmt.Sprintf("%s.%s", string(r), loRaAllianceJoinEUIDomain)
+}
+
+func HandleJoinRequest(ctx context.Context, tlsConfig *tls.Config, netID types.NetID, joinEUI, devEUI types.EUI64, req *ttnpb.JoinRequest, fqdn string) (*ttnpb.JoinResponse, error) {
+	dlSettings, err := lorawan.MarshalDLSettings(req.DownlinkSettings)
+	if err != nil {
+		return nil, ErrJoinReq.WithCause(err)
+	}
+
+	var cfList []byte
+	if req.CFList != nil {
+		cfList, err = lorawan.MarshalCFList(*req.CFList)
+		if err != nil {
+			return nil, ErrJoinReq.WithCause(err)
+		}
+	}
+
+	buf := &bytes.Buffer{}
+	if err := json.NewEncoder(buf).Encode(&JoinReq{
+		NsJsMessageHeader: NsJsMessageHeader{
+			MessageHeader: MessageHeader{
+				ProtocolVersion: "1.0",
+				MessageType:     MessageTypeJoinReq,
+				// TODO: Add TransactionID if necessary
+				// TODO: Add SenderToken if necessary
+			},
+			SenderID:   NetID(netID),
+			ReceiverID: EUI64(joinEUI),
+			SenderNSID: NetID(netID),
+		},
+		MACVersion: MACVersion(req.SelectedMACVersion),
+		PHYPayload: Buffer(req.RawPayload),
+		DevEUI:     EUI64(devEUI),
+		DevAddr:    DevAddr(req.DevAddr),
+		DLSettings: Buffer(dlSettings),
+		RxDelay:    req.RxDelay,
+		CFList:     Buffer(cfList),
+	}); err != nil {
+		return nil, err
+	}
+
+	scheme := "http"
+	var cl http.Client
+	if tlsConfig != nil {
+		scheme = "https"
+		cl.Transport = &http.Transport{
+			TLSClientConfig: tlsConfig,
+		}
+	}
+
+	res, err := cl.Post(fmt.Sprintf("%s://%s", scheme, fqdn), "application/json", buf)
+	if err != nil {
+		return nil, ErrJoinReq.WithCause(err)
+	}
+	defer res.Body.Close()
+
+	ans := &JoinAns{}
+	if err := json.NewDecoder(res.Body).Decode(ans); err != nil {
+		return nil, ErrJoinReq.WithCause(err)
+	}
+	if ans.Result.ResultCode != ResultSuccess {
+		return nil, ErrJoinReq
+	}
+
+	if ans.FNwkSIntKey != nil && ans.NwkSKey != nil && !reflect.DeepEqual(ans.FNwkSIntKey, ans.NwkSKey) {
```

Comment:
Is this spec?

Can't you just require the concerning session keys based on the MAC version?